File: core/src/test/java/com/nfthub/core/service/CategoryServiceTest.java
package com.nfthub.core.service;

import com.nfthub.core.entity.Category;
import com.nfthub.core.exception.AlreadyExistException;
import com.nfthub.core.exception.NotFoundException;
import com.nfthub.core.repository.CategoryRepository;
import com.nfthub.core.response.CategoryResponse;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@SpringBootTest
@Transactional
public class CategoryServiceTest {
    @Autowired
    private CategoryService categoryService;
    @Autowired
    private CategoryRepository categoryRepository;

    @Test
    @DisplayName("getCategoriesRes - 카테고리 목록 조회")
    public void getCategoriesRes() {
        categoryService.createCategory("test");
        categoryService.createCategory("test2");
        List<CategoryResponse> categories = categoryService.getCategoriesRes();
        Assertions.assertEquals(2, categories.size());
    }

    @Test
    @DisplayName("getCegoryResByNameOrThrow - 이름으로 카테고리 조회")
    public void getCategoryResByNameOrThrow() {
        CategoryResponse savedCategoryRes = categoryService.createCategory("test");
        CategoryResponse categoryRes = categoryService.getCategoryResByNameOrThrow("test");
        Assertions.assertEquals(savedCategoryRes.getId(), categoryRes.getId());
    }

    @Test
    @DisplayName("createCategory - 카테고리 생성")
    public void createCategory() {
        CategoryResponse createdCategoryRes = categoryService.createCategory("test");
        Category category = categoryRepository.findById(createdCategoryRes.getId()).get();
        Assertions.assertEquals(createdCategoryRes.getId(), category.getId());
    }

    @Test
    @DisplayName("createCateogory - 이미 같은 이름이 존재하면 throw")
    public void createCategoryAlreadyExist() {
        categoryService.createCategory("test");
        Assertions.assertThrows(AlreadyExistException.class, () -> categoryService.createCategory("test"));
    }

    @Test
    @DisplayName("updateCateogory - 카테고리 수정")
    public void updateCategory() {
        CategoryResponse createdCategoryRes = categoryService.createCategory("test");
        CategoryResponse updatedCategoryRes = categoryService.updateCategory(createdCategoryRes.getId(), "test2");
        Category category = categoryRepository.findById(updatedCategoryRes.getId()).get();
        Assertions.assertEquals("test2", category.getName());
    }

    @Test
    @DisplayName("updateCateogory - 이미 같은 이름이 존재하면 throw")
    public void updateCategoryAlreadyExist() {
        CategoryResponse res = categoryService.createCategory("test");
        categoryService.createCategory("test2");
        Assertions.assertThrows(AlreadyExistException.class, () -> categoryService.updateCategory(res.getId(), "test2"));
    }

    @Test
    @DisplayName("deleteCategory - 카테고리 삭제")
    public void deleteCategory() {
        CategoryResponse createdCategoryRes = categoryService.createCategory("test");
        categoryService.deleteCategory(createdCategoryRes.getId());
        Assertions.assertFalse(categoryRepository.findById(createdCategoryRes.getId()).isPresent());
    }

    @Test
    @DisplayName("deleteCategory - 존재하지 않는 카테고리 삭제시 throw")
    public void deleteCategoryNotExist() {
        Assertions.assertThrows(NotFoundException.class, () -> categoryService.deleteCategory(1L));
    }
}

File: src/test/kotlin/com/nfthub/api/service/CategoryServiceTest.kt
package com.nfthub.api.service

import com.nfthub.api.controller.ConflictException
import com.nfthub.api.controller.NotFoundException
import com.nfthub.api.dto.CategoryResponse
import com.nfthub.api.entity.Category
import com.nfthub.api.repository.CategoryRepository
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.data.repository.findByIdOrNull
import javax.transaction.Transactional
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

@SpringBootTest
@Transactional
class CategoryServiceTest(
    @Autowired val categoryService: CategoryService,
    @Autowired val categoryRepository: CategoryRepository
) {
    @Test
    fun `getCategoryResponses`() {
        categoryRepository.save(Category(name = "category1"))
        categoryRepository.save(Category(name = "category2"))
        categoryRepository.save(Category(name = "category3"))
        val categories = categoryService.getCategoryResponses()
        assertEquals(3, categories.size)
    }

    @Test
    fun `getCategoryOrThrow - 카테고리가 없으면 throw`() {
        assertThrows<NotFoundException> {
            categoryService.getCategoryOrThrow(1L)
        }
    }

    @Test
    fun `getCategoryResponse`() {
        val savedCategory = categoryRepository.save(Category(name = "category1"))
        val categoryResponse = categoryService.getCategoryResponse(savedCategory.id)
        assertEquals(
            categoryResponse, CategoryResponse(
                id = savedCategory.id, savedCategory.name
            )
        )
    }

    @Test
    fun `getCategoryResponse - 카테고리가 없으면 throw`() {
        assertThrows<NotFoundException> {
            categoryService.getCategoryResponse(1L)
        }
    }

    @Test
    fun `createCategory`() {
        val createdCategory = categoryService.createCategory("categoryName")
        val category = categoryRepository.findByIdOrNull(createdCategory.id)
        assertNotNull(category)
    }

    @Test
    fun `createCategory - 이미 존재하는 카테고리 이름이면, throw`() {
        categoryService.createCategory("category")
        assertThrows<ConflictException> {
            categoryService.createCategory("category")
        }
    }

    @Test
    fun `updateCategoryName`() {
        val savedCategory = categoryRepository.save(Category(name = "c"))
        categoryService.updateCategoryName(savedCategory.id, "category")
        val category = categoryRepository.findByIdOrNull(savedCategory.id)
        assertEquals(category?.name, "category")
    }

    @Test
    fun `updateCategoryName - 이미 존재하는 카테고리 이름이면, throw`() {
        val savedCategory = categoryRepository.save(Category(name = "c"))
        categoryService.updateCategoryName(savedCategory.id, "category")
        assertThrows<ConflictException> {
            categoryService.updateCategoryName(savedCategory.id, "category")
        }
    }

    @Test
    fun `deleteCategory`() {
        val savedCategory = categoryRepository.save(Category(name = "c"))
        categoryService.deleteCategory(savedCategory.id)
        val category = categoryRepository.findByIdOrNull(savedCategory.id)
        assertNull(category)
    }

    @Test
    fun `deleteCategory - 카테고리가 없으면 throw`() {
        assertThrows<NotFoundException> {
            categoryService.deleteCategory(1L)
        }
    }
}